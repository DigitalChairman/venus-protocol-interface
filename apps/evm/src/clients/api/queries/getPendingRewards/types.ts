import type BigNumber from 'bignumber.js';

import type { PoolLens, Prime, VaiVault, VenusLens, XvsVault } from 'libs/contracts';
import type { ChainId, MerklDistribution, Token } from 'types';
import type { Address } from 'viem';

export interface GetPendingRewardsInput {
  tokens: Token[];
  isolatedPoolComptrollerAddresses: string[];
  xvsVestingVaultPoolCount: number;
  accountAddress: Address;
  poolLensContract: PoolLens;
  xvsVaultContract: XvsVault;
  vaiVaultContract?: VaiVault;
  venusLensContract?: VenusLens;
  primeContract?: Prime;
  legacyPoolComptrollerContractAddress?: string;
  chainId: ChainId;
  merklCampaigns: Record<string, MerklDistribution[]>; // maps Asset -> Merkl campaigns
}

interface PendingRewardEntry {
  vTokenAddress: string;
  amountMantissa: BigNumber;
}

export type PendingInternalRewardSummary = {
  type: 'legacyPool' | 'isolatedPool';
  poolComptrollerAddress: string;
  distributorAddress: string;
  rewardTokenAddress: string;
  totalRewards: BigNumber;
  pendingRewards: PendingRewardEntry[];
};

export type PendingExternalRewardSummary = {
  type: 'external';
  appName: string;
  campaignId: string;
  campaignName: string;
  claimUrl: string;
  distributorAddress: string;
  rewardTokenAddress: string;
  totalRewards: BigNumber;
  pendingRewards: PendingRewardEntry[];
};

export interface GetPendingRewardsOutput {
  pendingRewardGroups: PendingRewardGroup[];
}

export interface IsolatedPoolPendingReward {
  rewardToken: Token;
  rewardAmountMantissa: BigNumber;
  rewardAmountCents: BigNumber | undefined;
  vTokenAddressesWithPendingReward: string[];
  rewardsDistributorAddress: string;
}

export interface ExternalPendingReward {
  campaignName: string;
  appName: string;
  claimUrl: string;
  rewardToken: Token;
  rewardAmountMantissa: BigNumber;
  rewardAmountCents: BigNumber | undefined;
  rewardsDistributorAddress: string;
}

export interface IsolatedPoolPendingRewardGroup {
  type: 'isolatedPool';
  comptrollerAddress: string;
  pendingRewards: IsolatedPoolPendingReward[];
}

export interface LegacyPoolPendingRewardGroup {
  type: 'legacyPool';
  comptrollerAddress: string;
  rewardToken: Token;
  rewardAmountMantissa: BigNumber;
  rewardAmountCents: BigNumber | undefined;
  vTokenAddressesWithPendingReward: string[];
}

export interface ExternalPendingRewardGroup {
  type: 'external';
  claimUrl: string;
  campaignName: string;
  appName: string;
  pendingRewards: ExternalPendingReward[];
}

export interface VaultPendingRewardGroup {
  type: 'vault';
  isDisabled: boolean;
  stakedToken: Token;
  rewardToken: Token;
  rewardAmountMantissa: BigNumber;
  rewardAmountCents: BigNumber | undefined;
}

export interface XvsVestingVaultPendingRewardGroup {
  type: 'xvsVestingVault';
  isDisabled: boolean;
  poolIndex: number;
  stakedToken: Token;
  rewardToken: Token;
  rewardAmountMantissa: BigNumber;
  rewardAmountCents: BigNumber | undefined;
}

export interface PrimePendingReward {
  rewardToken: Token;
  rewardAmountMantissa: BigNumber;
  rewardAmountCents: BigNumber | undefined;
}

export interface PrimePendingRewardGroup {
  type: 'prime';
  isDisabled: boolean;
  vTokenAddressesWithPendingReward: string[];
  pendingRewards: PrimePendingReward[];
}

export type PendingRewardGroup =
  | LegacyPoolPendingRewardGroup
  | IsolatedPoolPendingRewardGroup
  | VaultPendingRewardGroup
  | XvsVestingVaultPendingRewardGroup
  | PrimePendingRewardGroup
  | ExternalPendingRewardGroup;
