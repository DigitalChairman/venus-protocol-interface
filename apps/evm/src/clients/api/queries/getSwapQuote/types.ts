export interface ZeroXQuoteResponse {
  sellAmount: string;
  buyAmount: string;
  estimatedPriceImpact: string;
  price: string;
  data: string;
  allowanceTarget: string;
}
