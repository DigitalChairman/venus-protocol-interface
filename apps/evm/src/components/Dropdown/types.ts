import type { CardProps } from 'components/Card';

export interface DropdownOption<TValue extends string | number = string | number> {
  value: TValue;
  label:
    | string
    | React.ReactNode
    | ((context: { isRenderedInButton: boolean }) => string | React.ReactNode);
}

export interface DropdownProps<TValue extends string | number = string | number>
  extends Omit<CardProps, 'onChange' | 'children'> {
  children: ({ isDropdownOpened }: { isDropdownOpened: boolean }) => React.ReactNode;
  value?: TValue;
  options: DropdownOption<TValue>[];
  onChange?: (newValue: DropdownOption<TValue>['value']) => void;
  onBlur?: React.FocusEventHandler<HTMLInputElement>;
  name?: string;
  'data-testid'?: string;
  placeLabelToLeft?: boolean;
  label?: string;
  className?: string;
  buttonClassName?: string;
  size?: 'small' | 'medium' | 'large';
  variant?: 'primary' | 'secondary' | 'tertiary' | 'quaternary';
  menuTitle?: string;
  menuPosition?: 'left' | 'right';
}
