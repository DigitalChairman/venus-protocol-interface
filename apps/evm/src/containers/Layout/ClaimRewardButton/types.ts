import type BigNumber from 'bignumber.js';

import type { Claim } from 'clients/api';
import type { Token } from 'types';

export interface PendingReward {
  rewardToken: Token;
  rewardAmountMantissa: BigNumber;
  rewardAmountCents: BigNumber | undefined;
}

interface RewardsGroup {
  id: string;
  name: string;
  pendingRewards: PendingReward[];
  warningMessage?: string | React.ReactNode;
}

export interface InternalRewardsGroup extends RewardsGroup {
  isChecked: boolean;
  claims: Claim[];
  isDisabled?: boolean;
}

export interface ExternalRewardsGroup extends RewardsGroup {
  claimUrl: string;
  campaignName: string;
  appName: string;
}

export type Group = InternalRewardsGroup | ExternalRewardsGroup;
