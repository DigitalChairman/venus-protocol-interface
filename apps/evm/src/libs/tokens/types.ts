import type { ChainId } from '@venusprotocol/chains';
import type { Token, TokenAction } from 'types';

export type TokenMapping = {
  [chainId in ChainId]: Token[];
};

export type DisabledTokenActionMapping = {
  [chainId in ChainId]: DisabledTokenAction[];
};

export interface DisabledTokenAction {
  address: string;
  disabledActions: TokenAction[];
}
