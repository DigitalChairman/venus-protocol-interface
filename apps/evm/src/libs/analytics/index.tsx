export * from './useAnalytics';
export * from './AnalyticProvider';
