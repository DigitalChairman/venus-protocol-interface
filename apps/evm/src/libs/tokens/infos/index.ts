export * from './commonTokens';
export * from './pancakeSwapTokens';
export * from './vTokens';
