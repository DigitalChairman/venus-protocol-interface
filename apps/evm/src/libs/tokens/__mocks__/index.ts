import tokens from '__mocks__/models/tokens';
import type { ChainId } from 'types';
import { vTokenAssetsPerChainId } from '../infos/vTokens';

export * from 'libs/tokens/infos';

export const useGetToken = vi.fn(({ symbol }) => tokens.find(token => token.symbol === symbol));

export const useGetTokens = vi.fn(() => tokens);
export const getTokens = vi.fn(() => tokens);

export const getVTokenAsset = vi.fn(
  ({ vTokenAddress, chainId }: { vTokenAddress: string; chainId: ChainId }) =>
    vTokenAssetsPerChainId[chainId][vTokenAddress.toLowerCase()],
);

export const useGetSwapTokens = vi.fn(() => tokens);
export const getSwapTokens = vi.fn(() => tokens);

export const getDisabledTokenActions = vi.fn(() => []);
