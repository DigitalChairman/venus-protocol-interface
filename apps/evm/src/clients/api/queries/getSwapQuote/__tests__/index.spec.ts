import { bnb, xvs } from '__mocks__/models/tokens';
import { VError } from 'libs/errors';
import { ChainId } from 'types';
import type Vi from 'vitest';
import { type GetSwapQuoteInput, getSwapQuote } from '../';
import type { ZeroXQuoteResponse } from '../types';

const fakeParams: GetSwapQuoteInput = {
  apiUrl: 'fake-api-url',
  chainId: ChainId.BSC_TESTNET,
  fromToken: xvs,
  toToken: bnb,
  direction: 'exactAmountIn',
  fromTokenAmountTokens: '1',
};

const fakeQuoteResponse: ZeroXQuoteResponse = {
  sellAmount: '10000000000000000',
  buyAmount: '20000000000000000',
  estimatedPriceImpact: '0.5',
  price: '20000000000000000',
  data: 'fake-transaction-data',
  allowanceTarget: 'fake-allowance-target',
};

describe('getSwapQuote', () => {
  it.each([
    { direction: 'exactAmountIn', fromTokenAmountTokens: undefined },
    { direction: 'exactAmountIn', fromTokenAmountTokens: '0' },
    { direction: 'exactAmountOut', toTokenAmountTokens: undefined },
    { direction: 'exactAmountOut', toTokenAmountTokens: '0' },
  ] as Partial<GetSwapQuoteInput>[])(
    'returns an empty response when input is invalid: %s',
    async input => {
      const result = await getSwapQuote({ ...fakeParams, ...input });

      expect(result).toStrictEqual({
        swap: undefined,
        error: undefined,
      });
    },
  );

  it('throws unexpected error if request to 0x API fails', async () => {
    (fetch as Vi.Mock).mockImplementationOnce(() => {
      throw new Error('fake error');
    });

    try {
      await getSwapQuote(fakeParams);

      throw new Error('getSwapQuote should have thrown an error but did not');
    } catch (error: any) {
      expect(error).toBeInstanceOf(VError);
      expect(error.code).toBe('somethingWentWrong');
    }
  });

  it('returns INSUFFICIENT_LIQUIDITY error if requests succeeds but returns no swap', async () => {
    (fetch as Vi.Mock).mockImplementationOnce(() => ({
      json: () => ({
        code: 1,
      }),
    }));

    const result = await getSwapQuote(fakeParams);

    expect(result).toStrictEqual({
      swap: undefined,
      error: 'INSUFFICIENT_LIQUIDITY',
    });
  });

  it('returns "exact amount in" swap in the correct format on success', async () => {
    (fetch as Vi.Mock).mockImplementationOnce(() => ({
      json: () => fakeQuoteResponse,
    }));

    const result = await getSwapQuote({
      ...fakeParams,
      direction: 'exactAmountIn',
      fromTokenAmountTokens: '1',
    });

    expect(result).toMatchSnapshot();
  });

  it('returns "exact amount out" swap in the correct format on success', async () => {
    (fetch as Vi.Mock).mockImplementationOnce(() => ({
      json: () => fakeQuoteResponse,
    }));

    const result = await getSwapQuote({
      ...fakeParams,
      direction: 'exactAmountOut',
      toTokenAmountTokens: '1',
    });

    expect(result).toMatchSnapshot();
  });
});
