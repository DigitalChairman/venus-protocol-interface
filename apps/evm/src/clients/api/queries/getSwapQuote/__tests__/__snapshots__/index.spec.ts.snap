// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getSwapQuote > returns "exact amount in" swap in the correct format on success 1`] = `
{
  "error": undefined,
  "swap": {
    "direction": "exactAmountIn",
    "exchangeRate": "20000000000000000",
    "expectedToTokenAmountReceivedMantissa": "20000000000000000",
    "fromToken": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "fromTokenAmountSoldMantissa": "10000000000000000",
    "minimumToTokenAmountReceivedMantissa": "20000000000000000",
    "priceImpactPercentage": 0.5,
    "routePath": [],
    "toToken": {
      "address": "0x0000000000000000000000000000000000000000",
      "asset": "/src/libs/tokens/img/bnb.svg",
      "decimals": 18,
      "isNative": true,
      "symbol": "BNB",
    },
    "transactionData": "fake-transaction-data",
  },
}
`;

exports[`getSwapQuote > returns "exact amount out" swap in the correct format on success 1`] = `
{
  "error": undefined,
  "swap": {
    "direction": "exactAmountOut",
    "exchangeRate": "20000000000000000",
    "expectedFromTokenAmountSoldMantissa": "10000000000000000",
    "fromToken": {
      "address": "0xB9e0E753630434d7863528cc73CB7AC638a7c8ff",
      "asset": "/src/libs/tokens/img/xvs.svg",
      "decimals": 18,
      "symbol": "XVS",
    },
    "maximumFromTokenAmountSoldMantissa": "10000000000000000",
    "priceImpactPercentage": 0.5,
    "routePath": [],
    "toToken": {
      "address": "0x0000000000000000000000000000000000000000",
      "asset": "/src/libs/tokens/img/bnb.svg",
      "decimals": 18,
      "isNative": true,
      "symbol": "BNB",
    },
    "toTokenAmountReceivedMantissa": "20000000000000000",
    "transactionData": "fake-transaction-data",
  },
}
`;
